service/policy/db/attributes.go

```diff
@@ -55,227 +54,85 @@ func attributesValuesProtojson(valuesJSON []byte, attrFqn sql.NullString) ([]*po
-				avt.Field("id") + " AS av_id," +
-				"JSON_AGG(JSON_BUILD_OBJECT(" +
-				"'id', " + smT.Field("id") + "," +
-				"'actions', " + smT.Field("actions") + "," +
-				constructMetadata(smT.Name(), true) +
-				"'subject_condition_set', JSON_BUILD_OBJECT(" +
-				"'id', " + scsT.Field("id") + "," +
-				constructMetadata(scsT.Name(), true) +
-				"'subject_sets', " + scsT.Field("condition") +
-				")" +
-				")) AS sub_maps_arr " +
-				", inner_fqns.fqn AS fqn " +
-				"FROM " + smT.Name() + " " +
-				"LEFT JOIN " + avt.Name() + " ON " + smT.Field("attribute_value_id") + " = " + avt.Field("id") + " " +
-				"LEFT JOIN " + scsT.Name() + " ON " + smT.Field("subject_condition_set_id") + " = " + scsT.Field("id") + " " +
-				"INNER JOIN " + fqnt.Name() + " AS inner_fqns ON " + avt.Field("id") + " = inner_fqns.value_id " +
-				"WHERE inner_fqns.fqn = '" + opts.withOneValueByFqn + "' " +
-				"AND " + avt.Field("active") + " = true " +
-				"GROUP BY " + avt.Field("id") + ", inner_fqns.fqn " +
-				") AS val_sm_fqn_join ON " + "avt.id" + " = val_sm_fqn_join.av_id " +
-				"AND " + "avt.id" + " = " + fqnt.Field("value_id"),
-			)
-	}
-	g := []string{t.Field("id"), nt.Field("name")}
-
-	if opts.withFqn {
-		g = append(g, fqnt.Field("fqn"))
-	}
-
-	return sb.GroupBy(g...)
-}
-
-func attributesHydrateItem(row pgx.Row, opts attributesSelectOptions, logger *logger.Logger) (*policy.Attribute, error) {
-	if opts.withKeyAccessGrants {
-		opts.withAttributeValues = true
-	}
-
-	var (
-		id            string
-		name          string
-		rule          string
-		metadataJSON  []byte
-		namespaceID   string
-		active        bool
-		namespaceName string
-		valuesJSON    []byte
-		valuesOrder   []string
-		grants        []byte
-		fqn           sql.NullString
-	)
-
-	fields := []interface{}{&id, &name, &rule, &metadataJSON, &namespaceID, &active, &namespaceName}
-	shouldGetValues := opts.withAttributeValues || opts.withOneValueByFqn != "" || opts.withFqn
-	if shouldGetValues {
-		fields = append(fields, &valuesOrder, &valuesJSON)
-	}
-	if opts.withKeyAccessGrants {
-		fields = append(fields, &grants)
-	}
-	if opts.withFqn {
-		fields = append(fields, &fqn)
-	}
-
-	err := row.Scan(fields...)
-	if err != nil {
-		return nil, db.WrapIfKnownInvalidQueryErr(err)
-	}
-
-	m := &common.Metadata{}
-	if metadataJSON != nil {
-		if err := protojson.Unmarshal(metadataJSON, m); err != nil {
+func hydrateAttribute(row *attributeQueryRow, logger *logger.Logger) (*policy.Attribute, error) {
+	metadata := &common.Metadata{}
+	if row.metadataJSON != nil {
+		if err := protojson.Unmarshal(row.metadataJSON, metadata); err != nil {
 			logger.Error("could not unmarshal metadata", slog.String("error", err.Error()))
 			return nil, err
 		}
 	}
-	var v []*policy.Value
-	if valuesJSON != nil {
-		v, err = attributesValuesProtojson(valuesJSON, fqn)
+
+	var values []*policy.Value
+	if row.valuesJSON != nil {
+		v, err := attributesValuesProtojson(row.valuesJSON, row.fqn)
```

Comment:
The reason this function took in the attribute FQN originally was in case the selected values did not contain the value FQNs and we needed to build them dynamically for the response. In other words, it was to work around squirrel limitations.

I think now we can ensure we have each value FQN populated by the SQL query itself and update this `attributesValuesProtojson` to no longer take in the attribute FQN.

Reply:
This has been moved to the SQL